source/hash_table.c

```diff
@@ -0,0 +1,582 @@
+            // item
+            rv = AWS_ERROR_HASHTBL_ITEM_NOT_FOUND;
+            break;
+        }
+
+        probe_idx++;
+    } while(1);
+
+    *p_entry = entry;
+    if (p_probe_idx) {
+        if (probe_cap < probe_idx) {
+            probe_idx = probe_cap;
+        }
+        *p_probe_idx = probe_idx;
+    }
+
+    return rv;
+}
+
+int aws_common_hash_table_find(struct aws_common_hash_table *map,
+    const void *key, struct aws_common_hash_element **pElem) {
+    struct hash_table_state *state = map->pImpl;
+    uint64_t hash_code = hash_for(state, key);
+    struct hash_table_entry *entry;
+
+    int rv = find_entry(state, hash_code, key, &entry, NULL);
+
+    if (rv == AWS_ERROR_SUCCESS) {
+        *pElem = &entry->element;
+    }
+
+    return rv;
+}
+
+/*
+ * Attempts to find a home for the given entry. Returns after doing nothing if entry was not occupied.
+ */
+static struct hash_table_entry *emplace_item(struct hash_table_state *state, struct hash_table_entry entry, int probe_idx) {
+    struct hash_table_entry *initial_placement = NULL;
+
+    while (entry.hash_code) {
+        int index = (entry.hash_code + probe_idx) & state->mask;
+        struct hash_table_entry *victim = &state->slots[index];
+
+        int victim_probe_idx = (index - victim->hash_code) & state->mask;
+
+        if (!victim->hash_code || victim_probe_idx < probe_idx) {
+            if (!initial_placement) {
+                initial_placement = victim;
+            }
+
+            struct hash_table_entry tmp = *victim;
+            *victim = entry;
+            entry = tmp;
+
+            probe_idx = victim_probe_idx + 1;
+        } else {
+            probe_idx++;
+        }
+    }
+
+    return initial_placement;
+}
+
+static int expand_table(struct aws_common_hash_table *map) {
+    struct hash_table_state *old_state = map->pImpl;
+    struct hash_table_state template = *old_state;
+
+    update_template_size(&template, template.size * 2);
+
+    struct hash_table_state *new_state = alloc_state(&template);
+    if (!new_state) {
+        return AWS_ERROR_OOM;
+    }
+
+    for (int i = 0; i < old_state->size; i++) {
+        struct hash_table_entry entry = old_state->slots[i];
+        if (entry.hash_code) {
+            // We can directly emplace since we know we won't put the same item twice
+            emplace_item(new_state, entry, 0);
+        }
+    }
+
+    map->pImpl = new_state;
+    aws_mem_release(new_state->alloc, old_state);
+
+    return AWS_ERROR_SUCCESS;
+}
+
+int aws_common_hash_table_create(struct aws_common_hash_table *map,
+    const void *key, struct aws_common_hash_element **pElem, int *was_created
+) {
+    struct hash_table_state *state = map->pImpl;
+    uint64_t hash_code = hash_for(state, key);
+    struct hash_table_entry *entry;
+    int probe_idx;
+    int ignored;
+    if (!was_created) {
+        was_created = &ignored;
```

Comment:
should ignored have been initialized at this point?

Reply:
Nope. We're writing into it, not reading from it.